ffi_how_to.md

```diff
@@ -0,0 +1,115 @@
+This document aims to create a detailed work plan for shipping the `pre-built-ffi` workstream.
+The high level approach for this has already been well documented at https://hackmd.io/@mvdan/Hy7iK0TEY.
+
+### Changes to https://github.com/filecoin-project/filecoin-ffi
+
+Going ahead, except for cases where users want to build `filecoin-ffi` from source, `filecoin-ffi` will essentially act as a "wrapper" that forwards all API calls to the corresponding `prebuilt-ffi` module. Since a `prebuilt-ffi` module contains platform-dependent assets (such as static C libraries), APIs in `filecoin-ffi` will need to delegate calls to the "platform-specific" `prebuilt-ffi` module.
+
+Currently, `filecoin-ffi` supports multiple (GOOS + GOARCH) combinations. Therefore, we must modify every public API in `filecoin-ffi` to have a variant for each (GOOS + GOARCH) combination. These variants will delegate the call to the corresponding platform-specific `prebuilt-ffi` module instead of calling into the CGO bindings as they do today (except when users build `filecoin-ffi` from source).
+
+Furthermore, it is crucial to ensure complete backward compatibility for current users of `filecoin-ffi`, including those who opt to compile `filecoin-ffi` from the source.
+
+This objective can be achieved using Go build tags. 
+For illustration, consider the existing public `ffi.Hash()` API in the `bls.go` file of `filecoin-ffi`. Below is an outline of the necessary variants for this API:
+
+```go
+prebuilt_bls_darwin_arm64.go
+//go:build cgo && darwin && arm64 && !ffi_source
+// +build cgo,darwin,arm64
+// +build !ffi_source
+package ffi
+import (
+	prebuilt "fil.org/prebuilt-ffi-darwin-arm64"
+    // When building for darwin/arm64, Go tooling automatically selects this file due to the specified build 
+    // tags. It then fetches the "prebuilt-ffi-darwin-arm64" module from the module proxy, using the version 
+    // specified in the `go.mod` file.
+)
+
+// Hash computes the digest of a message
+func Hash(message Message) Digest {
+	return prebuilt.Hash(message)
+}
+```
+
+```go
+prebuilt_bls_linux_amd64.go
+//go:build cgo && linux && amd64 && !ffi_source
+// +build cgo,linux,amd64
+// +build !ffi_source
+package ffi
+import (
+	prebuilt "fil.org/prebuilt-ffi-linux-amd64"
+)
+
+// Hash computes the digest of a message
+func Hash(message Message) Digest {
+	return prebuilt.Hash(message)
+}
+```
+
+```go
+// Build from source
+bls__source.go
+//go:build ffi_source
+// +build ffi_source
+// Same code as we have today
+```
+
+And so on and so forth for each (GOOS + GOARCH) combination for each of those files containing public APIs that currently call into the CGO bindings.
+
+The logical next question to ask is how are the `prebuilt-ffi-{GOOS}-{GOARCH}` modules mentioned here created and where are they hosted?
+
+### Building and publishing the prebuit-ffi modules + CI changes
```

Comment:
I think it would also be good to be clear about the authoring experience.  Lets say I have something I want to add cgo/bls.  Do I just add it to `bls.go`?  Will there be codegen to create `prebuilt_bls_darwin_arm64.go`?

Reply:
Great question. I hadn't thought about this. This one is a bit hard to get right in a seamless way. One option is to add a layer of indirection by introducing a new `filecoin-ffi-prebuilt` repo that end users depend on and keeping `filecoin-ffi` completely unchanged so devs can continue hacking on it like they do now. The e2e flow then would look like:

1) Dev changes `bls.go` in `filecoin-ffi`
2) Cuts a new release of `filecoin-ffi`
3) CI takes over and :
    - Clones this new `filecoin-ffi` release 
    - Creates and persists the prebuilt-assets for the combinations of {GOARCH}-{GOOS} we need to support to release assets
    - Uses codegen to generate the `prebuilt-*` files for the `filecoin-ffi-prebuilt` repo
    - Updates the go.mod for the `filecoin-ffi-prebuilt` repo to depend on the `prebuilt-{GOARCH}-{GOOS}` repos
    - Releases the `filecoin-ffi-prebuilt` repo
    
    
End users then depend on `filecoin-ffi-prebuilt` instead of `filecoin-ffi`.

But, as you can see, this is a fair bit of work.